src/net_processing.cpp

```diff
@@ -3195,6 +3210,140 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+    // children instead of the real one provided by the honest peer.
+    const auto cpfp_candidates_same_peer{m_orphanage.GetChildrenFromSamePeer(ptx, nodeid)};
+
+    // These children should be sorted from most newest to oldest. In the (probably uncommon) case
+    // of children that replace each other, this helps us accept the highest feerate (probably the
+    // most recent) one efficiently.
+    for (const auto& child : cpfp_candidates_same_peer) {
+        Package maybe_cpfp_package{ptx, child};
+        if (!m_recent_rejects_reconsiderable.contains(GetPackageHash(maybe_cpfp_package))) {
+            tx_orphan = child;
+            orphan_sender = nodeid;
+            break;
+        }
+    }
+
+    // If no suitable candidate from the same peer is found, also try children that were provided by
+    // a different peer. This is useful because sometimes multiple peers announce both transactions
+    // to us, and we happen to download them from different peers (we wouldn't have known that these
+    // 2 transactions are related). We still want to find 1p1c packages then.
+    //
+    // If we start tracking all announcers of orphans, we can restrict this logic to parent + child
+    // pairs in which both were provided by the same peer, i.e. delete this step.
+    if (!tx_orphan) {
+        const auto cpfp_candidates_different_peer{m_orphanage.GetChildrenFromDifferentPeer(ptx, nodeid)};
+
+        // Find the first 1p1c that hasn't already been rejected. We randomize the order to not
+        // create a bias that attackers can use to delay package acceptance.
+        //
+        // Create a random permutation of the indices.
+        std::vector<size_t> tx_indices(cpfp_candidates_different_peer.size());
+        std::iota(tx_indices.begin(), tx_indices.end(), 0);
+        Shuffle(tx_indices.begin(), tx_indices.end(), m_rng);
+
+        for (const auto index : tx_indices) {
+            // If we already tried a package and failed for any reason, the combined hash was
+            // cached in m_recent_rejects_reconsiderable.
+            Package maybe_cpfp_package{ptx, cpfp_candidates_different_peer.at(index).first};
+            if (!m_recent_rejects_reconsiderable.contains(GetPackageHash(maybe_cpfp_package))) {
+                tx_orphan = cpfp_candidates_different_peer.at(index).first;
+                orphan_sender = cpfp_candidates_different_peer.at(index).second;
+                break;
+            }
+        }
+    }
+
+    if (tx_orphan) {
+        const Package package_1p1c{ptx, tx_orphan};
+        const std::vector<NodeId> senders{nodeid, orphan_sender};
+        const auto package_result{ProcessNewPackage(m_chainman.ActiveChainstate(), m_mempool, package_1p1c, /*test_accept=*/false, /*client_maxfeerate=*/std::nullopt)};
+        LogDebug(BCLog::TXPACKAGES, "package evaluation for parent %s (wtxid=%s) + child %s (wtxid=%s) in orphanage: %s\n",
```

Comment:
FWIW: I added logging here to see what number of parents and given by different peers than the orphan. After a day of running it's 0 out of 162. 

Is there a reason to think this should be a common pattern?

Reply:
If a parent+child are relayed back to back across the network, then I would expect there to be times when we download the transactions out of order (eg because we send a request for the parent to a different peer than the request for the child).

Reply:
I guess it'll likely be something like:

1) peer A sends child, child is put in orphanage and parent tx request queued(but not yet sent)
2) peer B has slightly out of date feefilter for your node, sends INV for parent

Reply:
fwiw it's definitely not common, but I do see them occasionally and have a handful of acceptances of packages from 2 different peers

Reply:
> though I'm not convinced this really will make a difference

Reply:
> I guess it'll likely be something like:
> 
> 1. peer A sends child, child is put in orphanage and parent tx request queued(but not yet sent)
> 2. peer B has slightly out of date feefilter for your node, sends INV for parent

What I'd expect to happen is that we get inv's for parent+child from both Peer A and Peer B, and we happen to request the parent from A and the child from B, but the child arrives first -- it's put in the orphanage, and since a request for the parent is already in flight to A, we don't send another request for the parent to B.

Reply:
> Is the suggestion to drop this and only try same-peer packages?

If this is hitting actual usage, no, because it's coded and I actually reviewed it, and tests seem to cover it. I'm unsure if in BIP331-like world it would be used since it would be receiver-driven, but that might be thinking too far ahead.

> (In this scenario I guess I'm assuming no feefilter, or that the feefilter value is slightly stale.)

Reply:
Ok, marking this as resolved